nle/env/base.py

```diff
@@ -511,6 +532,32 @@ def get_seeds(self):
         """
         return self.env.get_current_seeds()
 
+    def tty_render(self, obs, keys=("tty_chars", "tty_colors"), print_guides=True):
+        tty_chars_index = self._observation_keys.index(keys[0])
+        tty_colors_index = self._observation_keys.index(keys[1])
+        tty_chars = obs[tty_chars_index]
+        tty_colors = obs[tty_colors_index]
+        H, W = tty_chars.shape
+        rendering = ""
+        col_index_str = " " + COLORS[1] + "".join([str(i % 10) for i in range(W)])
+        if print_guides:
+            rendering += col_index_str + "\n"
+        for i in range(H):
+            line = []
+            for j in range(W):
+                char = tty_chars[i, j]
+                color = tty_colors[i, j]
+                line.append(COLORS[color] + chr(char))
+            row_index_str = COLORS[1] + str(i % 10)
+            if print_guides:
+                rendering += row_index_str + "".join(line) + row_index_str + "\n"
```

Comment:
in general this string additions can get very slow because in Python string are immutable, and + is a binary operator.
so `"some long" + "strings" + "can be" + "expensive" + "to make" `

is evaluated as:
```
("some long" + "strings")  # all copied to "some long strings"
("some long string"+ "can be")  # copied to "some long strings can be"
("some long strings can be"  + "expensive")  ... # etc O(N^2)
```

Thats why the "Pythonic" thing is to collect the strings in a list and use `.join` method as you have above, which means the allocation is only done once (O(N))

Reply:
TBF you have done the line.join for for the rows, so N in this case is only like 24 rather than 80, but its still an order of magnitude.